src/dune_pkg/opam_solver.ml

```diff
@@ -424,27 +419,30 @@ let opam_package_to_lock_file_pkg ~repo ~local_packages opam_package =
       Loc.none, Package_name.of_string (OpamPackage.Name.to_string name))
   in
   let build_command =
+    let subst_step =
+      OpamFile.OPAM.substs opam_file
+      |> List.map ~f:(fun x ->
+        let x = OpamFilename.Base.to_string x in
+        let input = String_with_vars.make_text Loc.none (x ^ ".in") in
+        let output = String_with_vars.make_text Loc.none x in
+        Action.Substitute (input, output))
+      |> progn_opt
+    in
     let patch_step =
       (* CR-someday alizter: Patches don't take into account filters that are present. For
          now we take them all. *)
-      match
-        OpamFile.OPAM.patches opam_file
-        |> List.map ~f:(fun (x, _) ->
-          Action.Patch
-            (String_with_vars.make_text Loc.none (OpamFilename.Base.to_string x)))
-      with
-      | [] -> None
-      | [ x ] -> Some x
-      | xs -> Some (Action.Progn xs)
+      OpamFile.OPAM.patches opam_file
+      |> List.map ~f:(fun (x, _) ->
+        Action.Patch (String_with_vars.make_text Loc.none (OpamFilename.Base.to_string x)))
+      |> progn_opt
+    in
+    let build_step =
+      opam_commands_to_actions opam_package (OpamFile.OPAM.build opam_file) |> progn_opt
     in
-    opam_commands_to_action opam_package (OpamFile.OPAM.build opam_file)
-    |> Option.map ~f:(fun action ->
-      match patch_step with
-      | None -> action
-      | Some patch_step -> Action.Progn [ patch_step; action ])
+    List.filter_opt [ subst_step; patch_step; build_step ] |> progn_opt
```

Comment:
You could leave all the `{subst,patch,build}_step` as lists and use `List.concat`. That way you could inline `progn_opt` which is a bit of an ugly function to keep around.

Reply:
Sure thing.